src/ArraySetops.chpl

```diff
@@ -336,4 +337,404 @@ module ArraySetops
+              }
+            }
+          }
+        }
+      }
+
+      sortAndUpdateStats();
+
+      proc updateRetLocales(){
+        var sum: int = 0;
+        var retLoc: int = 0;
+
+        for i in 0..<(len){
+
+          const size : int = aSize[i] + bSize[i];
+
+          if((sum + size) <= segs[Locales[retLoc].id]){ // Data fits on current retLoc
+            returnLocId[i] = Locales[retLoc].id;
+            returnSize[i] = size;
+            sum += size;
+          }else if (( (sum + size) == segs[Locales[retLoc].id] + 1) && (a[aIndex[i]] == b[bIndex[i]])){  // Data is only one too big for current retLoc.
+            returnLocId[i] = Locales[retLoc].id;
+            returnSize[i] = size;
+            sum += size;
+          }else if(sum < segs[Locales[retLoc].id]){
+            returnLocId[i] = Locales[retLoc].id;
+            returnSize[i] = segs[Locales[retLoc].id] - sum;
+            needsSplit[i] = true;
+            sum += size - segs[Locales[retLoc].id];
+            retLoc += 1;
+          }else{
+            retLoc += 1;
+            returnLocId[i] = Locales[retLoc].id;
+            returnSize[i] = size;
+            sum = size;
+          }
+        }
+      }
+
+      updateRetLocales();
+
+      //  Find the location to split off the min K of the two arrays.
+      //  Array 2 will have fewer lookups.
+      proc findMinKLocations(k: int, const ref arry1: [?D] ?t, const ref arry2: [] t, low1:  int, high1:  int, low2:  int, high2: int): (int, int, t){
+
+        const maxIterations : int = floor(log(k)):int;
+
+        var rangeLow1: int = low1;
+        var rangeHigh1 : int = high1;
+
+        var guessIndex1: int = floor((high1 + low1)/2):int;
+        var guessIndex2: int = low2 + k - (guessIndex1  - low1);
+        var val1: int = arry1[guessIndex1];
+        var val2: int = arry2[guessIndex2];
+
+        proc update(){
+          if( val1 > val2){
+            rangeHigh1 = guessIndex1;
+          } else{
+            rangeLow1 = guessIndex1;
+          }
+
+          guessIndex1 = floor((rangeHigh1 + rangeLow1)/2): int ;
+          guessIndex2 = low2 + k - (guessIndex1  - low1);
+
+          val1 = arry1[guessIndex1];
+          val2 = arry2[guessIndex2];
+        }
+
+        for i in 0..maxIterations{
+          if(val1 == val2){
+            return (guessIndex1, guessIndex2, max(val1, val2));
+          }if(val1 > val2 && arry1[guessIndex1-1] <= val2){
+            return (guessIndex1, guessIndex2, max(val1, val2));
+          }if(val1 < val2 && arry1[guessIndex1+1] >= val2){
+            return (guessIndex1, guessIndex2, max(val1, val2));
+          }else{
+            update();
+          }
+        }
+
+        return (guessIndex1, guessIndex2, max(val1, val2));
+      }
+
+      //  Determine split points for cases when the segment needs to be divided between locales.
+      coforall loc in Locales with (const ref a, const ref b, ref len){
+        on loc {
+          for i in 0..<(len){
```

Comment:
Any ideas how to turn this into a coforall?  The computations are all independent.  It's just the writing step that would cause a problem if performed concurrently.